File: daily/2025/1931_Painting_a_Grid_With_Three_Different_Colors.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int colorTheGrid(int m, int n) {
        int mod = 1e9 + 7;
        int type = pow(3, m);
        vector<vector<int>> color_map(type, vector<int>(m));
        for (int c = 0; c < type; c++){
            int curr = c;
            for (int i = 0; i < m; i++){
                color_map[c][i] = curr % 3;
                curr /= 3;
            }
        }
        vector<int> valid;
        for (int i = 0; i < type; i++){
            bool v = true;
            for (int j = 1; j < m; j++){
                if (color_map[i][j-1] == color_map[i][j]){
                    v = false;
                    break;
                }
            }
            if (v) valid.push_back(i);
        }
        vector<vector<int>> valid_pair(type);
        for (int i : valid){
            for (int j : valid){
                bool v = true;
                for (int c = 0; c < m; c++){
                    if (color_map[i][c] == color_map[j][c]){
                        v = false;
                        break;
                    }
                }
                if (v) valid_pair[i].push_back(j);
            }
        }
        vector<vector<int>> dp(type, vector<int>(n, 0));
        for (int i : valid) dp[i][0] = 1;
        for (int j = 0; j < n - 1; j++){
            for (int i = 0; i < type; i++){
                if (dp[i][j]){
                    for (int next : valid_pair[i]){
                        dp[next][j+1] = (dp[next][j+1] + dp[i][j]) % mod;
                    }
                }
            }
        }
        int ans = 0;
        for (int i = 0; i < type; i++){
            ans = (ans + dp[i][n-1]) % mod;
        }
        return ans;
    }
};

int main(){
    // int m = 1;
    // int n = 1;
    // int m = 1;
    // int n = 2;
    int m = 5;
    int n = 5;
    Solution S;
    cout << S.colorTheGrid(m, n) << endl;
    return 0;
}

File: daily_python/1931_Painting_a_Grid_With_Three_Different_Colors.py
import os
import sys
from typing import List

class Solution:
    def colorTheGrid(self, m: int, n: int) -> int:
        t = 3 ** m
        valid = {}
        mod = int(1e9 + 7)
        for color in range(t):
            idx = color
            temp = []
            for c in range(m):
                temp.append(color % 3)
                color //= 3
            v = True
            for c in range(m-1):
                if temp[c] == temp[c+1]:
                    v = False
                    break
            if v:
                valid[idx] = temp
        pair = {}
        for x, x_c in valid.items():
            pair[x] = []
            for y, y_c in valid.items():
                v = True
                for c in range(m):
                    if x_c[c] == y_c[c]:
                        v = False
                        break
                if v:
                    pair[x].append(y)
        dp = [[0] * n for _ in range(t)]
        for i in valid.keys():
            dp[i][0] = 1
        for  j in range(n-1):
            for i in range(t):
                if dp[i][j]:
                    for next in pair[i]:
                        dp[next][j+1] = (dp[next][j+1] + dp[i][j]) % mod
        ans = 0
        for i in range(t):
            ans = (ans + dp[i][n-1]) % mod
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            m = int(lines[0].split('=')[1].strip()[:-1])
            n = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((m, n))
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for m, n in dataset:
        results.append(solution.colorTheGrid(m, n))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')